director/director.go

```diff
@@ -288,9 +299,79 @@ func redirectToCache(ginCtx *gin.Context) {
 	if err != nil {
 		log.Errorf("Failed to get depth attribute for the redirecting request to %q, with best match namespace prefix %q", reqPath, namespaceAd.Path)
 	}
+
+	// Stat origins and caches for object availability
+	// For origins, we only want ones with the object
+	// For caches, we still list all in the response but turn down priorities for ones that don't have the object
+	originAdsWObject := []server_structs.ServerAd{}
+	// An array to keep track of object availability of each caches
+	cachesAvailabilityMap := make(map[string]bool, len(cacheAds))
+
+	if skipStat {
+		originAdsWObject = originAds
+		for _, cAd := range cacheAds {
+			cachesAvailabilityMap[cAd.URL.String()] = true
+		}
+	} else {
+		// Query Origins and check if the object exists on the server
+		q := NewObjectStat()
+		st := config.NewServerType()
+		st.SetList([]config.ServerType{config.OriginType, config.CacheType})
+		// Set max response to all available origin and cache servers to ensure we stat against origins
+		// if no cache server has the file
+		// TODO: come back and re-evaluate if we need this many responses and potential origin/cache
+		// server performance issue out of this
+		maxRes := len(cacheAds) + len(originAds)
+		qr := q.Query(context.Background(), reqPath, st, 1, maxRes,
+			withOriginAds(originAds), withCacheAds(cacheAds), WithToken(reqParams.Get("authz")))
+		log.Debugf("Stat result for %s: %s", reqPath, qr.String())
+
+		// For successful response, we got a list of URLs to access the object.
+		// We will use the host of the object url to match the URL field in originAds and cacheAds
+		if qr.Status == querySuccessful {
+			for _, obj := range qr.Objects {
+				serverHost := obj.URL.Host
+				for _, oAd := range originAds {
+					if oAd.URL.Host == serverHost {
+						originAdsWObject = append(originAdsWObject, oAd)
+					}
+				}
+				for _, cAd := range cacheAds {
+					if cAd.URL.Host == serverHost {
+						cachesAvailabilityMap[cAd.URL.String()] = true
+					}
+				}
+			}
+		} else if qr.Status == queryFailed {
+			if qr.ErrorType != queryInsufficientResErr {
+				ginCtx.JSON(http.StatusInternalServerError, server_structs.SimpleApiResp{
+					Status: server_structs.RespFailed,
+					Msg:    fmt.Sprintf("Failed to query origins with error %s: %s", string(qr.ErrorType), qr.Msg),
+				})
+				return
+			}
+			// For denied servers, append them to availableOriginAds
+			// The qr.DeniedServers is a list of AuthURLs of servers that respond with 403
+			// Here, we need to match against the AuthURL field of originAds
+			for _, ds := range qr.DeniedServers {
```

Comment:
I don't follow why we'd be getting denied from these servers. Don't we pass an auth token along with the stat in line 325?

Reply:
This is in case the user doesn't know if the origin is protected or not. The future plan is to return 403 with issuer URLs to the denied servers, if all servers denied the stat request, so that the client and automatically invoke OAuth flow

Reply:
But shouldn't the Director know whether access for a particular resource requires authorization? If the user tries to stat `/protected/foo` and the Director knows `/protected` requires a token, why not just return a 403 to the user instead of performing the actual stat?

My gut says we should handle "the director knows an origin is protected and the user didn't provide a stat token" differently from "the director knows an origin is protected and the user provided an access token, but the token wasn't accepted by the origin".

Reply:
Yes! That'll be addressed in the issue https://github.com/PelicanPlatform/pelican/issues/1309 I think by then we can say 1. If accessing object from a protected namespace, the director returns 403 straight. 2. If token is given but the all responded stat gave 403, director should respond 403 as well